core/market/decentralized/src/negotiation/requestor.rs

```diff
@@ -43,7 +54,132 @@ impl RequestorNegotiationEngine {
         &self,
         subscription_id: &SubscriptionId,
     ) -> Result<(), NegotiationError> {
-        // TODO: Implement
+        self.notifier.stop_notifying(subscription_id).await;
+
+        // We can ignore error, if removing events failed, because they will be never
+        // queried again and don't collide with other subscriptions.
+        let _ = self
+            .db
+            .as_dao::<EventsDao>()
+            .remove_requestor_events(subscription_id)
+            .await
+            .map_err(|e| {
+                log::warn!(
+                    "Failed to remove events related to subscription [{}].",
+                    subscription_id
+                )
+            });
+        // TODO: We could remove all resources related to Proposals.
         Ok(())
     }
+
+    pub async fn query_events(
+        &self,
+        subscription_id: &SubscriptionId,
+        timeout: f32,
+        max_events: Option<i32>,
+    ) -> Result<Vec<RequestorEvent>, QueryEventsError> {
+        let mut timeout = Duration::from_millis((1000.0 * timeout) as u64);
```

Comment:
```suggestion
        let mut timeout = Duration::from_secs_f32(timeout);
```

https://doc.rust-lang.org/std/time/struct.Duration.html#method.from_secs_f32

Reply:
Thanks to this I found potential bug. Conversion to u64 could do something strange in case of negative timeouts